### ISSUE TITLE
Epoch Timing Data Migration Plan

### ISSUE THREAD

**Issue by topocount** [coordinape/coordinape#1850]
## Context

As a part of creating the new epoch forms, and needed backends to support
the increased degrees of configurability, we're going to migrate all epochs
to the new backend data structures as a part of the merge process.

To execute this cutover cleanly, this migration script must exist as a hasura
migration that executes on merge. This will minimize down-time, but it will
require nontrivial testing to ensure the script is robust and works with
all existing prod data.

## Logic

All existing repeating epochs can be migrated to the custom repeating epoch
`repeat_data` config. This data has the following shape:

```ts
type CustomRepeatData = {
  type: 'custom';
  frequency: number;
  frequency_unit: 'days' | 'weeks' | 'months';
  start_date: zStringISODateUTC;
  end_date: zStringISODateUTC;
};
```

This maps quite well to our exiting configuration range of a duration of
X days, with repeat intervals of `weekly`, and `monthly`, where the epoch
cannot exceed 7 days for `weekly` and 28 days for `monthly`.

> **Note**
> These older, extant durations are not timezone-aware, and when converted to
> start and end dates on the backend will either truncate or extend in time
> if a change to or from Daylight Savings occurs in within the duration.

## Proposed migration sequence

In plain english, we want to convert this duration to a start and end date
and convert the frequencies according to:

**weekly repeat**

```json
{
  "frequency": 1,
  "frequency_unit": "weeks"
}
```

**monthly repeat**

```json
{
  "frequency": 1,
  "frequency_unit": "months"
}
```

As long as the start and end datetimes max out at `1 week` or `1 month` by [`luxon`](https://moment.github.io/luxon/#/math)'s
calculation (we can truncate epoch end dates that do exceed these respective limits), the configurations
will be valid.

[postgres's date time calculation](https://www.postgresql.org/docs/current/functions-datetime.html) utilities will stand in for luxon for this script.

### technical notes

PostgresSQL is capable of internally executing this migration through its
[jsonb support](https://www.postgresql.org/docs/current/datatype-json.html#JSONB-SUBSCRIPTING)

This work can be completed on top of the https://github.com/coordinape/coordinape/tree/update-epoch-handler branch

**Comment by levity** [comment]
does updating the cronjob for epochs have to wait until this migration is complete? or could it go in a PR stacked on top of the one that contains the migration?

**Comment by crabsinger** [comment]
> As long as the start and end datetimes max out at 1 week or 1 month by [luxon](https://moment.github.io/luxon/#/math)'s
calculation (we can truncate epoch end dates that do exceed these respective limits), the configurations
will be valid.

If the duration is greater than the repeat interval, truncate it so it fits

**Comment by crabsinger** [comment]
Before doing any work to handle the long duration invariant, I am checking if we have any prod data that matches this invariant:

### Find any weekly repeating epoches with a duration longer than 7 days:
```sql
SELECT * from epoches
WHERE repeat = '1'
AND end_date - start_date > INTERVAL '7 days'
AND ended = false
LIMIT 5
```

no results

### Find any monthly repeating epoches with a duration longer than 1 month
```sql
SELECT * from epoches
WHERE repeat = '2'
AND end_date - start_date > INTERVAL '28 days'
LIMIT 5
```
no results

So I think its safe to ignore these duration invariants @topocount

**Comment by crabsinger** [comment]
i feel confident there are no edge cases on this for weekly repeating epochs in prod.

**Comment by crabsinger** [comment]
for monthly: edge case would be months repeating with duration greater than 28 days, which we have none of.

**Comment by topocount** [comment]
@crabsinger now that the "custom" JSON includes an explicit `duration` and `duration_unit`. This can just be set to

```JSON
{
  "duration": "<value from days column>",
  "duration_unit": "days"
}
```

I'm happy to make this change if you'd prefer

**Comment by topocount** [comment]
> does updating the cronjob for epochs have to wait until this migration is complete? or could it go in a PR stacked on top of the one that contains the migration?

@levity I'm going to stack the cron work on top of this. I think it makes the most sense to squash and merge them together, actually, so the existing cron task doesn't produce any non-migrated epochs while we merge and deploy the new cron logic.
